src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateClone = { ...state };
+  const stateHistory = [];
+
+  for (const obj of actions) {
+    if (obj.type === 'addProperties') {
+      Object.assign(stateClone, obj.extraData);
+      stateHistory.push({ ...stateClone });
+    }
+
+    if (obj.type === 'removeProperties') {
+      for (const delObj of obj.keysToRemove) {
+        delete stateClone[delObj];
+      }
+      stateHistory.push({ ...stateClone });
+    }
+
+    if (obj.type === 'clear') {
+      for (const key in stateClone) {
+        if (Object.prototype.hasOwnProperty.call(stateClone, key)) {
+          delete stateClone[key];
```

Comment:
When handling the 'clear' action, there's no need to loop through the keys of the state clone to delete them individually. You can simply assign an empty object to `currentState` to achieve the same result more efficiently.